File: Leetcode/dfs_recursion_backtracking/733_Flood Fill/Solution.py
class Solution(object):
    def floodFill(self, image, sr, sc, newColor):
        """
        :type image: List[List[int]]
        :type sr: int
        :type sc: int
        :type newColor: int
        :rtype: List[List[int]]
        """
        if image[sr][sc]==newColor:
            return image
        oldColor = image[sr][sc]
        def dfs(sr, sc):
            if (not (0 <= sr < len(image) and 0 <= sc < len(image[sr]))):
                return
            if image[sr][sc]!=oldColor:
                return
            image[sr][sc]=newColor
            [dfs(sr + x, sc + y) for (x, y) in ((0, 1), (1, 0), (0, -1), (-1, 0))]
        dfs(sr, sc)
        return image

File: Leetcode/dfs_recursion_backtracking/733_Flood Fill/Solution.java
class Solution {
    public int[][] floodFill(int[][] image, int sr, int sc, int newColor) {
        if( image[sr][sc]==newColor )
            return image;
        dfs(image, image[sr][sc], newColor, sr, sc);
        return image;
    }
        private void dfs(int[][] image, int oldColor, int newColor,
                        int i, int j){
            if( i<0 || j<0 || i>=image.length || j>=image[i].length)
                return;
            if( image[i][j]!=oldColor || image[i][j]==newColor )
                return;
            image[i][j] = newColor;
            dfs(image, oldColor, newColor, i+1, j);
            dfs(image, oldColor, newColor, i-1, j);
            dfs(image, oldColor, newColor, i, j+1);
            dfs(image, oldColor, newColor, i, j-1);
        }
}
